### PR TITLE
Introduce @EnabledIf execution condition

### LINKED ISSUE #219: Introduce predefined @Disabled*/@Enabled* variants
## Overview

I just happened to have a test only failing on MS Windows machines and thought about JUnit Jupiter.

I could write a custom extension implementing `ContainerExecutionCondition` and `TestExecutionCondition`, but having a predefined `@Disabled` annotation such as `@DisabledIfWindows` or `@Disabled(condition = IsWindowsEvaluator.class)` to extend disabling feature might be handy.

## Related Issues

- #1247

## Deliverables

- [x] system property matches a pattern
- [x] environment variable matches a pattern
- ❌  current day of week matches one of the supplied enums
  - The team decided not to publish `@EnabledOnDayOfWeek` and `@DisabledOnDayOfWeek`.
- [x] current OS matches one of the supplied enums
- [x] current JRE version matches one of the supplied enums
- [x] Check for user configuration errors. For example, if a user declares `@EnabledOnOs({})`, that should result in an exception instead of a disabled result.
- [x] Introduce integration tests for each condition.
- [x] Write JavaDoc for each annotation.
- [x] Document in User Guide.
- [x] Document in Release Notes.

### PR DESCRIPTION
**Pull request by sormuras**
## Overview

Supply a JavaScript script controlling whether the annotated container or test is executed.

Closes #219

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by sormuras** [comment]
**TODO** Add `String reason()` back to the annotation.

**Comment by ttddyy** [comment]
Hi there,

This is pretty interesting.
If `@EnabledIf` is extensible as proposed in #1257 (e.g: `engine` parameter) and evaluation implementation(`EnabledIfCondition`) is extendable, users(or other frameworks) can simply use this annotation and register a script engine to evaluate the expression, instead of implementing own `@EnabledIf` annotation and `ExecutionCondition`.

For example, `@EnabledIf` in spring5 can just be a composed annotation that defines `engine="SpEL"` with other spring specific parameters. Then, extends `EnabledIfCondition` from jupiter to add logic to parse the spel expression.
Even more, it seems `ScriptEngineManager` uses `ServiceLoader`, so, SpEL parsing logic(currently `EnabledIfCondition` in spring) can be an implementation of `ScriptEngine` and automatically be registered.
(sounds bit over-engineering though 😅 )  cc @sbrannen

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+	/**
+	 * The lines of the script controlling the <em>enabled</em> or <em>disabled</em> state.
+	 */
+	String[] value();
+
+	/**
+	 * Short name of the {@link javax.script.ScriptEngine ScriptEngine} to use.
+	 *
+	 * <p>Use Oracle Nashorn as the default JavaScript ScriptEngine.
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: "is used by default"?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * <li>{@link Bind#JUPITER_TAGS jupiterTags} - All tags as a {@code Set<String>}.</li>
+ * <li>{@link Bind#JUPITER_DISPLAY_NAME jupiterDisplayName} - Display name as a {@code String}.</li>
+ * <li>{@link Bind#JUPITER_UNIQUE_ID jupiterUniqueId} - Unique ID as a {@code String}.</li>
+ * <li>{@link Bind#JUPITER_CONFIGURATION_PARAMETER jupiterConfigurationParameter} - Configuration parameter accessor.</li>
+ * </ul>
+ *
+ * @since 5.1
+ * @see org.junit.jupiter.api.extension.ExecutionCondition
+ * @see org.junit.jupiter.api.extension.ConditionEvaluationResult#enabled(String)
+ * @see org.junit.jupiter.api.extension.ConditionEvaluationResult#disabled(String)
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: Add `@see ScriptEngine`?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	 * The lines of the script controlling the <em>enabled</em> or <em>disabled</em> state.
+	 */
+	String[] value();
+
+	/**
+	 * Short name of the {@link javax.script.ScriptEngine ScriptEngine} to use.
+	 *
+	 * <p>Use Oracle Nashorn as the default JavaScript ScriptEngine.
+	 *
+	 * Until Java SE 7, JDKs shipped with a JavaScript scripting engine based
```
Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: Missing `<p>`?

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/EnabledIfCondition.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+		return evaluate(annotation, scriptEngine, script);
+	}
+
+	private ConditionEvaluationResult evaluate(EnabledIf annotation, ScriptEngine scriptEngine, String script) {
+		Object result;
+		try {
+			result = scriptEngine.eval(script);
+		}
+		catch (ScriptException e) {
+			logger.warn(e, () -> "Evaluation of @EnabledIf script failed, disabling execution");
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/EnabledIfCondition.java]: Shouldn't we let the test fail when evaluating the script fails?

Comment (sormuras) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/EnabledIfCondition.java]: Good point. Throw an `JUnitException`?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java (hunk 2 above)
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: All documentation polishing is targetet to **5.1-RC1** ;)

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/EnabledIfCondition.java (hunk 4 above)
Comment (marcphilipp) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/EnabledIfCondition.java]: Yep! 👍

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java (hunk 2 above)
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: Done.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java (hunk 3 above)
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: Done.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java (hunk 1 above)
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/EnabledIf.java]: Done.

**Pull request merged**